pkg/reconciler/pipeline/controller.go

```diff
@@ -534,6 +534,21 @@ func needsUpdate(old, new *dfv1.Pipeline) bool {
 	if !equality.Semantic.DeepEqual(old.Finalizers, new.Finalizers) {
 		return true
 	}
+
+	oldAnnotations := old.GetAnnotations()
+	newAnnotations := new.GetAnnotations()
+	if oldAnnotations != nil && newAnnotations == nil {
+		return true
+	}
+	if oldAnnotations == nil && newAnnotations != nil {
+		return true
+	}
+
+	for k, v := range oldAnnotations {
```

Comment:
The logic we want is to compare the annotations with key starting with `numaflow.numaproj.io/`, if there's difference, return true. Difference could be number of key/value pairs are different, same key has different value, etc.